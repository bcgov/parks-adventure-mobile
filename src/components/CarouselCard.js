import React from 'react'
import PropTypes from 'prop-types'
import { useTheme } from 'react-native-paper'
import MCIcon from 'react-native-vector-icons/MaterialCommunityIcons'
import defaultParkImage from '../../assets/defaultParkImage.jpg'
import {
  Carousel,
  CardCover,
  ParkTitle,
  CardContent,
  ContentLine,
  ParkDistance,
  CardBanner,
  AdvisoryText,
} from './CarouselCard.styles'

function CarouselCard({ onPress, onFavoritePress, park, distance }) {
  const theme = useTheme()
  const { title, imageUri, favorited, alerts, advisories } = park

  //  Format advisory banner message text
  const totalAdvisoryCount = alerts.length + advisories.length
  let advisoryMessage

  if (totalAdvisoryCount === 1 && alerts.length === 1) {
    advisoryMessage = alerts[0].Headline
  } else if (totalAdvisoryCount === 1 && advisories.length === 1) {
    advisoryMessage = advisories[0].Headline
  } else {
    if (alerts.length === 0) {
      advisoryMessage = `${advisories.length} Advisories`
    } else if (advisories.length === 0) {
      advisoryMessage = `${alerts.length} Alerts`
    } else {
      advisoryMessage = `${totalAdvisoryCount} Alerts/Advisories`
    }
  }

  return (
    <Carousel
      accessibilityLabel={'navigate to park details'}
      elevation={6}
      onPress={onPress}>
      <CardCover
        imageStyle={{ borderTopLeftRadius: 13, borderTopRightRadius: 13 }}
        defaultSource={defaultParkImage}
        source={imageUri ? { imageUri } : defaultParkImage}
        resizeMode="cover">
        {totalAdvisoryCount > 0 && (
          <CardBanner alert={alerts.length > 0}>
            <AdvisoryText numberOfLines={1} alert={alerts.length > 0}>
              {advisoryMessage}
            </AdvisoryText>
          </CardBanner>
        )}
      </CardCover>
      <CardContent>
        <ContentLine>
          <ParkTitle numberOfLines={3}>{title}</ParkTitle>
          <MCIcon
            name={favorited ? 'heart' : 'heart-outline'}
            size={20}
            color={theme.colors.secondary500}
            onPress={onFavoritePress}
            accessibilityLabel="favorite park"
          />
        </ContentLine>
        <ParkDistance>{`${distance}km Away`}</ParkDistance>
      </CardContent>
    </Carousel>
  )
}

CarouselCard.propTypes = {
  onPress: PropTypes.func.isRequired,
  onFavoritePress: PropTypes.func.isRequired,
  distance: PropTypes.string.isRequired,
  park: PropTypes.shape({
    title: PropTypes.string.isRequired,
    imageUri: PropTypes.string,
    favorited: PropTypes.bool,
    alerts: PropTypes.arrayOf(
      PropTypes.shape({
        Headline: PropTypes.string.isRequired,
      })
    ),
    advisories: PropTypes.arrayOf(
      PropTypes.shape({
        Headline: PropTypes.string.isRequired,
      })
    ),
  }).isRequired,
}

export default CarouselCard
