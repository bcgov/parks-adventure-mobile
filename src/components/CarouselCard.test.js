import React from 'react'
import { render } from '@testing-library/react-native'
import renderSnapshot from 'react-test-renderer'
import CarouselCard from './CarouselCard'

test('Should render title and distance props', () => {
  const park = {
    title: 'Adventure Park',
    uri: 'https://picsum.photos/id/1043/5184/3456',
    distance: '297',
    favourited: false,
  }

  const { getByText } = render(<CarouselCard {...park} />)

  const title = getByText(park.title)
  const distance = getByText(`${park.distance}km Away`)

  expect(title).toBeDefined()
  expect(distance).toBeDefined()
})

test('renders with no uri and favourited set to true', () => {
  const park = {
    title: 'Adventure Park',
    distance: '297',
    favourited: true,
  }

  const tree = renderSnapshot.create(<CarouselCard {...park} />).toJSON()
  expect(tree).toMatchSnapshot()
})
