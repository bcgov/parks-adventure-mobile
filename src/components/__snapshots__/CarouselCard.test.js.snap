// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with no uri and favourited set to true 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": 13,
      "elevation": 6,
      "height": 220,
      "overflow": "hidden",
      "shadowColor": "rgba(0, 0, 0, 0.25)",
      "shadowOffset": Object {
        "height": 4,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 4,
      "width": 260,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#eeeeee",
            "height": 195,
            "overflow": "hidden",
          },
          Object {
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Array [
            Object {
              "height": 135,
            },
          ],
        ]
      }
    >
      <Image
        resizeMode="contain"
        siblings={
          Array [
            "Styled(withTheme(Card.Cover))",
            "Styled(Card.Content)",
          ]
        }
        source={1}
        style={
          Array [
            Object {
              "flex": 1,
              "height": undefined,
              "justifyContent": "flex-end",
              "padding": 16,
              "resizeMode": "cover",
              "width": undefined,
            },
            Object {
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingBottom": 16,
          },
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "marginTop": 15,
            },
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontWeight": "bold",
                  "width": "90%",
                },
              ],
            ]
          }
        >
          Adventure Park
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.26)",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï‹‘
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "bottom": 15,
                "fontSize": 12,
                "opacity": 0.6,
                "position": "absolute",
                "right": 15,
              },
            ],
          ]
        }
      >
        297km Away
      </Text>
    </View>
  </View>
</View>
`;
